llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+public:
+	GenericValue getOperandValueExplicit(const Value* V, ExecutionContext& EC)
+	{
+		return Interpreter::getOperandValue(const_cast<Value*>(V), EC);
+	}
+	GenericValue getOperandValue(const Value* V)
+	{
+		return getOperandValueExplicit(V, getLastStack());
+	}
+	void assignToMaps(const Value* V, const BitMask bitMask)
+	{
+		stronglyKnownBits[V] = bitMask;
+	}
+	void assignToMaps(const Value* V, const BitMask bitMask, GenericValue GV)
+	{
+		assignToMaps(V, bitMask);
+		getLastStack().Values[const_cast<Value*>(V)] = GV;
+	}
+	void assignToMaps(const Value* V, const Value* toAssign)
+	{
+		assignToMaps(V, getBitMask(toAssign), getOperandValue(toAssign)); 
+	}
+	void assignToMapsIfComputed(const Value* V, const Value* toAssign)
+	{
+		if (isValueComputed(toAssign))
+		{
+			assignToMaps(V, getBitMask(toAssign));
+		}
+	}
+	void removeFromMaps(const Value* V)
+	{
+		stronglyKnownBits.erase(V);
+		getLastStack().Values.erase(const_cast<Value*>(V));
+	}
+	
+	llvm::DenseMap<const llvm::Value*, BitMask> stronglyKnownBits;
+	BitMask computeStronglyKnownBits(const llvm::Instruction& I)
+	{
+		BitMask min = BitMask::ALL;
+
+		for (auto& op : I.operands())
+		{
+			min = BitMask(min & getBitMask(op));
+		}
+
+		if (min == BitMask::ALL)
+		{
+			//All operands fully known
+			return BitMask::ALL;
+		}
+
+		switch (I.getOpcode())
+		{
+			case Instruction::Load:
+			{
+				return BitMask::ALL;
+			}
+			case Instruction::Select:
+				// TODO(carlo): Select can also be specialised further
+			case Instruction::Add:
+			case Instruction::Sub:
+			case Instruction::Mul:
+			case Instruction::PtrToInt:
+			case Instruction::BitCast:
+			case Instruction::GetElementPtr:
+			{
+				//Those all work modulo N
+				return min;
+			}
+			case Instruction::And:
+			{
+				//If either is Constant && less than equal min -> ALL
+				if (const ConstantInt* CI = dyn_cast<ConstantInt>(I.getOperand(0)))
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+				if (const ConstantInt* CI = dyn_cast<ConstantInt>(I.getOperand(1)))
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+
+				[[clang::fallthrough]];
+			}
+			case Instruction::Or:
+			case Instruction::Xor:
+			{
+				//Those all work modulo 2^k
+				return min;
+			}
+			default:
+			{
+				//Some operands NOT fully known, means in general no information is conserved
+				return BitMask::NONE;
+			}
+		}
+	}
+	void setIncomingBB(const llvm::BasicBlock* from)
+	{
+		fromBB = from;
+	}
+	const llvm::BasicBlock* getIncomingBB() const
+	{
+		return fromBB;
+	}
+	enum VisitingPolicy {COMPUTE_VALUES, REMOVE_VALUES};
+	template <VisitingPolicy policy>
+	const llvm::BasicBlock* visitBasicBlock(const llvm::BasicBlock* BB, const llvm::BasicBlock* from=nullptr);
+	explicit PartialInterpreter(std::unique_ptr<llvm::Module> M)
+		: llvm::Interpreter(std::move(M), /*preExecute*/false)
+	{
+	}
+	BitMask getBitMask(const llvm::Value* V) const
```

Comment:
Should this be merged with computeStronglyKnownBits?

Reply:
Even though they are intertwined, I will review this